### PR TITLE
add KV cache dtype as explicit argument [PRI-347]

### LINKED ISSUE #1119: [PRI-347] Expose `kv_cache_dtype`
### Describe the workflow you want to enable

Currently the KV cache is quantized by default and cannot be used unquantized without hacking. 

### Describe your proposed solution

add `kv_cache_dtype` as an argument

### Describe alternatives you've considered, if relevant

_No response_

### Additional context

_No response_

### Impact

Low (Nice to have)

### PR DESCRIPTION
**Pull request by jmkuebler**
## Issue

Fixes PriorLabs/TabPFN#1119 and PriorLabs/TabPFN#631.  <br>Also this is an alternative to PriorLabs/TabPFN#1099

## Motivation and Context

`kv_cache_dtype`is a tradeoff decision and should be possible to control explicitly

---

## Public API Changes

- [ ] No Public API changes
- [X] Yes, Public API changes (Details below)

---

## How Has This Been Tested?

I fixed the unit tests to not rely on Monkeypatching anymore.  <br>I also ran the script shared in [https://github.com/PriorLabs/TabPFN/issues/631](<https://github.com/PriorLabs/TabPFN/issues/631>)

```python
import random
from sklearn.datasets import load_diabetes
from sklearn.model_selection import train_test_split
import numpy as np
import torch

from tabpfn import TabPFNRegressor

X, y = load_diabetes(return_X_y=True)
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.33, random_state=42
)


def _set_seeds() -> None:
    torch.manual_seed(0)
    np.random.seed(0)
    random.seed(0)


_set_seeds()
reg = TabPFNRegressor(fit_mode="low_memory", inference_precision=torch.float64)
reg.fit(X_train, y_train)
preds_no_cache = reg.predict(X_test)

reg = TabPFNRegressor(fit_mode="low_memory", inference_precision=torch.float64)
reg.fit(X_train, y_train)
preds_no_cache_repeat = reg.predict(X_test)

_set_seeds()
reg = TabPFNRegressor(fit_mode="fit_preprocessors", inference_precision=torch.float64)
reg.fit(X_train, y_train)
preds_cache_preproc = reg.predict(X_test)

_set_seeds()
reg = TabPFNRegressor(fit_mode="fit_with_cache", inference_precision=torch.float64, kv_cache_dtype="auto")
reg.fit(X_train, y_train)
preds_kv_cache_auto = reg.predict(X_test)

_set_seeds()
reg = TabPFNRegressor(fit_mode="fit_with_cache", inference_precision=torch.float64, kv_cache_dtype="int8")
reg.fit(X_train, y_train)
preds_kv_cache_int8 = reg.predict(X_test)


def _max_diff(a: np.ndarray, b: np.ndarray) -> float:
    return np.max(np.abs(a - b) / np.abs(a))


print("max relative diffs")
print("no_cache vs no_cache_repeat:", _max_diff(preds_no_cache, preds_no_cache_repeat))
print("no_cache vs cache_preproc:", _max_diff(preds_no_cache, preds_cache_preproc))
print("no_cache vs kv_cache (auto):", _max_diff(preds_no_cache, preds_kv_cache_auto))
print("no_cache vs kv_cache int8:", _max_diff(preds_no_cache, preds_kv_cache_int8))
```

Using `kv_cache_dtype="auto"` fixes the issue. If `int8`is used we now throw a warning, when someone forces the inference dtype to sth else.

```
/usr/lib/python3.12/contextlib.py:81: UserWarning: inference_precision forces a dtype but kv_cache_dtype is 'int8', so the KV cache is still stored as int8 regardless of the forced dtype. Pass kv_cache_dtype='auto' to keep the KV cache in the forced dtype.
  return func(*args, **kwds)
max relative diffs
no_cache vs no_cache_repeat: 0.0
no_cache vs cache_preproc: 0.0
no_cache vs kv_cache (auto): 0.0
```

---

## Checklist

- [X] The changes have been tested locally.
- [ ] Documentation has been updated (if the public API or usage changes).
- [ ] A changelog entry has been added (see `changelog/README.md`), or "no changelog needed" label requested.
- [X] The code follows the project's style guidelines.
- [X] I have considered the impact of these changes on the public API.

---

```
```

### CODE REVIEW

**Review by bejaeger — COMMENTED** [review]
Looks great thanks a lot!

I forgot to mention that not all our models currently supported kv cache quantization. So not only is the new `kv_cache_dtype` argument only relevant when fit_mode is `fit_with_cache` but also only if the model is `v3` (which is the only one that currently supports int8). This is not ideal and might be confusing if we have "int8" as the default. Should we have another "magic" option like "auto" that is "smallest" or "optimized" which will always pick the smallest supported kv cache dtype?

File: src/tabpfn/regressor.py [hunk 1]
```diff
@@ -243,6 +244,7 @@ def __init__(  # noqa: PLR0913
         ] = "fit_preprocessors",
         memory_saving_mode: MemorySavingMode = "auto",
         keep_cache_on_device: bool = True,
+        kv_cache_dtype: Literal["auto", "int8"] = "int8",
```
Comment (oscarkey) [inline src/tabpfn/regressor.py]: I'm slightly worried the parameter default value might cause us an issue if future models need a different value, because this default can't depend on the model version.

What if the default was `auto` and:

* cache_precision=auto + inference_precision=auto → cache dtype set to int8 for v3
* cache_precision=auto + inference_precision=specified as A → cache dtype is set to A
* cache_precision=specified as A + inference_precision=specified as B → cache dtype is set to A
  * This way we can also delete the warning, as the user is explicitly making the odd choice.

Comment (oscarkey) [inline src/tabpfn/regressor.py]: Will many end users change this? If not, we could put it in `inference_config.py`. It's still possible to override, but keeps the constructor simpler. Happy either way though.

Comment (oscarkey) [inline src/tabpfn/regressor.py]: Should the name be `cache_precision`?

* Consistent `keep_cache_on_device` and `inference_precision`
* Lots of our users might not be familiar with transformers and so not know what "kv" means.

Or maybe `cache_dtype`.

**Review by oscarkey — COMMENTED** [review]
Just a few thoughts! I'm on holiday tomorrow and monday, so happy with whatever's decided if you want to merge :)

File: src/tabpfn/regressor.py (hunk 1 above)
Comment (oscarkey) [reply src/tabpfn/regressor.py]: although this could be confused with the caching we already do of the input data in some of the fit modes, or any future cache we add, hmmm

Comment (bejaeger) [reply src/tabpfn/regressor.py]: Yeah we briefly discussed this actually. The problem is that there is more in the cache than just the kv cache and this is currently not quantized. So we opted for staying explicit here. I agree though that it's not ideal because very technical.
Maybe "hiding" it in the inference config is the way to go?

I'm indifferent on `precision` vs `dtype`. Harmonizing with `inference_precision` might be cleaner.

Comment (jmkuebler) [reply src/tabpfn/regressor.py]: kv_cache_precision it is

Comment (jmkuebler) [reply src/tabpfn/regressor.py]: the model version now has in the architecture what is supported and we resolve against that. Also made the argument optional. If `None` is provided we resolve to the defaults, otherwise we respect the user choice unless not supported, in whcih case we warn

File: src/tabpfn/architectures/interface.py [hunk 2]
```diff
@@ -250,6 +250,17 @@ def get_default_performance_options(self) -> PerformanceOptions:
...
         )
 
+    def get_supported_kv_cache_precisions(self) -> tuple[str, ...]:
+        """KV cache storage dtypes this architecture supports (``fit_with_cache``).
+
+        ``"auto"`` (keep the computed dtype) is always supported. Architectures
+        whose cache implements quantization override this to additionally list
+        e.g. ``"int8"``. Used to resolve the estimator's ``kv_cache_precision``:
+        ``None`` picks the most compact supported value, and an unsupported
+        request falls back to ``"auto"``.
```
Comment (bejaeger) [inline src/tabpfn/architectures/interface.py]: the last sentence seems like it doesn't belong here and rather should be in the estimator doc string?

File: src/tabpfn/architectures/tabpfn_v3.py [hunk 3]
```diff
@@ -1930,6 +1936,10 @@ def get_default_performance_options(self) -> PerformanceOptions:
             use_chunkwise_inference=True,
         )
 
+    @override
+    def get_supported_kv_cache_precisions(self) -> tuple[str, ...]:
+        return ("auto", "int8")
```
Comment (bejaeger) [inline src/tabpfn/architectures/tabpfn_v3.py:1941]: nit: could technically get this from the `TabPFNV3Cache` class but let's keep it like that for now.

File: src/tabpfn/inference.py [hunk 4]
```diff
@@ -894,7 +919,9 @@ def __init__(  # noqa: PLR0913
         )
 
         self.keep_cache_on_device = keep_cache_on_device
-        self.maybe_quantize_kv_cache = maybe_quantize_kv_cache
+        self.kv_cache_precision = resolve_kv_cache_precision(
+            kv_cache_precision, architecture=models[0]
+        )
```
Comment (bejaeger) [inline src/tabpfn/inference.py]: Can we make this a list instead for each model in `models` and below in build_cache use the model_index? Could otherwise bite us in the future. I'd say we then want to adjust the warning in the resolve function to only be thrown once?

**Review by bejaeger — APPROVED** [review]
Very nice! thanks for all the changes, looks very clean now.<br>Approving to avoid blocking you but please consider my other comments before merging.

File: src/tabpfn/architectures/interface.py (hunk 2 above)
Comment (jmkuebler) [reply src/tabpfn/architectures/interface.py]: thanks, removed that. It was unecessary to state in the docstring for what we would use that.

File: src/tabpfn/inference.py (hunk 4 above)
Comment (jmkuebler) [reply src/tabpfn/inference.py]: I moved the resolution now within each cache build, so it is per estimator.
